### DISCUSSION TITLE
Turnstile: error

### DISCUSSION THREAD

**Discussion by mike007-1** [seleniumbase/SeleniumBase#2877] [Q&A] [answered]
Hi,

I replaced the former turnstile click by "driver.uc_gui_handle_cf()", and I get this error: "module 'seleniumbase.config' has no attribute 'headless'" when the turnstile challenge is on.
I tried to launch it with and without headless mode, on Windows 11 and Ubuntu 20. Here is the code:

from seleniumbase import Driver
driver = Driver(uc=True,headless2=True)
driver.get(https://famousfoxes.com/tokenmarket)
driver.uc_gui_handle_cf()

**Comment by mdmintz** [comment, marked as answer]
UC Mode no longer works in Headless Mode, as it needs a GUI Display for `pyautogui`.

UC Mode now needs a special X11 virtual display to run on Linux, which is included with the `SB()` and `BaseCase` formats, but not the limited `Driver()` format.

Try the example:

```python
from seleniumbase import SB

with SB(uc=True, test=True) as sb:
    url = "https://www.virtualmanager.com/en/login"
    sb.uc_open_with_reconnect(url, 4)
    print(sb.get_page_title())
    sb.uc_gui_handle_cf()  # Ready if needed!
    print(sb.get_page_title())
    sb.assert_element('input[name*="email"]')
    sb.assert_element('input[name*="login"]')
    sb.set_messenger_theme(location="bottom_center")
    sb.post_message("SeleniumBase wasn't detected!")
```

**Comment by mike007-1** [comment]
If I understand, even without headless mode, the Driver() method does not solve anymore turnstile, and needs a GUI Display for pyautogui to launch the driver.uc_gui_handle_cf() ?
How to implement easily the GUI Display for pyautogui so that driver.uc_gui_handle_cf() works ?

**Reply by mdmintz** [reply]
A regular GUI display is fine. This only affects UC Mode on Linux servers without a GUI for Cloudflare specifically. The `Driver()` doesn't spin up `Xvfb` because there's no tearDown available to clean up the resources. With `SB()` / `BaseCase` formats, there's a context manager format, meaning both setUp and tearDown, and so resources can be cleaned up. Otherwise, this situation happens: https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues?q=is%3Aissue+is%3Aopen+memory (where uses don't properly clean up resources at the end, leading to memory leaks and system crashes when out-of-memory). There's a lot of SeleniumBase code to prevent those memory leaks via tearDown. So if you need the special X11 virtual display, then I suggest using the `SB()` or `BaseCase` [syntax formats](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md) unless there's a good reason to use the `Driver()` format.

**Comment by mdmintz** [comment]
Oh, did you mean this?

```python
File "SeleniumBase/seleniumbase/core/browser_launcher.py", line 552, in verify_pyautogui_has_a_headed_browser
    if sb_config.headless or sb_config.headless2:
       ^^^^^^^^^^^^^^^^^^
AttributeError: module 'seleniumbase.config' has no attribute 'headless'
```

Yeah, that's a bug with `Driver()` and `uc_gui_handle_cf()`. I'll fix it.

Next time, properly format your code, include the stack trace, and give your discussion a better title.

**Comment by mike007-1** [comment]
Exactly, that was my point.

`from seleniumbase import Driver`
`driver = Driver(uc=True)`
`driver.get(https://famousfoxes.com/tokenmarket)`
`driver.uc_gui_handle_cf()`

There is this error:
`  File "C:\Users\mike\AppData\Local\Programs\Python\Python312\Lib\site-packages\seleniumbase\core\browser_launcher.py", line 4025, in <lambda>`
`    lambda *args, **kwargs: uc_gui_handle_cf(`
`                            ^^^^^^^^^^^^^^^^^`
`  File "C:\Users\mike\AppData\Local\Programs\Python\Python312\Lib\site-packages\seleniumbase\core\browser_launcher.py", line 651, in uc_gui_handle_cf`
`    install_pyautogui_if_missing()`
`  File "C:\Users\mike\AppData\Local\Programs\Python\Python312\Lib\site-packages\seleniumbase\core\browser_launcher.py", line 559, in install_pyautogui_if_missing`
`    verify_pyautogui_has_a_headed_browser()`
`  File "C:\Users\mike\AppData\Local\Programs\Python\Python312\Lib\site-packages\seleniumbase\core\browser_launcher.py", line 552, in verify_pyautogui_has_a_headed_browser`
`    if sb_config.headless or sb_config.headless2:`
`       ^^^^^^^^^^^^^^^^^^`
`AttributeError: module 'seleniumbase.config' has no attribute 'headless'`

`module 'seleniumbase.config' has no attribute 'headless'`

**Reply by mdmintz** [reply]
This was resolved in `4.28.1` - https://github.com/seleniumbase/SeleniumBase/releases/tag/v4.28.1

**Reply by mike007-1** [reply]
Thank you, it works fine !

**Comment by mike007-1** [comment]
`from seleniumbase import Driver`
`driver = Driver(uc=True)`
`driver.get(https://famousfoxes.com/tokenmarket)`
`driver.uc_gui_handle_cf()`

On Linux Ubuntu 20 Xfce env, after the update, the driver.uc_gui_handle_cf() is aborted and returns:
`NOTE: You must install tkinter on Linux to use MouseInfo. Run the following: sudo apt-get install python3-tk python3-dev`

It works perfectly after installing the package ! Thanks again, you are a wizard !
